File: web/utils/tagName.ts
// Client-side twin of internal/graph/tagname (Go). The server is the authority —
// it normalises and rejects on every write path — but the input needs the same
// rules locally to tell the user *before* the request that `sys/` is reserved or
// that `foo//bar` is not a name, and to send the same canonical string the server
// would have derived anyway.
//
// Plain TS (no Vue imports) so the input component and unit tests share it.

export const TAG_SEPARATOR = '/'

// Machine-authored tags live under this prefix so they never compete with the
// user's own subjects in the tag surface. Mirrors tags.SystemPrefix.
export const SYSTEM_PREFIX = 'sys/'

export const isSystemTag = (name: string): boolean => name.startsWith(SYSTEM_PREFIX)

export interface TagNameResult {
  name?: string
  error?: string
}

// Canonical form: trimmed, lowercased, every `/`-separated segment trimmed.
// Lowercasing is what makes a tag a matching axis rather than prose — the graph
// stores names UNIQUE, so `Homelab` and `homelab` must converge.
export function normalizeTag(raw: string): TagNameResult {
  const trimmed = raw.trim()
  if (!trimmed) return { error: 'Tag name is empty.' }

  const segments = trimmed.toLowerCase().split(TAG_SEPARATOR).map((s) => s.trim())
  if (segments.some((s) => s === '')) {
    return { error: `"${raw}" has an empty segment. Nest tags as parent/child.` }
  }

  const name = segments.join(TAG_SEPARATOR)
  if (isSystemTag(name)) {
    return { error: `"${SYSTEM_PREFIX}" is reserved for system tags.` }
  }
  return { name }
}

File: internal/graph/tagname/tagname.go
// Package tagname normalises and validates tag names.
//
// Tags nest by convention: the name is a flat string whose `/` separators are
// read as a hierarchy at query time (`homelab/nas` is a child of `homelab`).
// That convention only holds if every write path agrees on what a name looks
// like, and there are two of them — tags.Add and the nodes chokepoint. The
// chokepoint cannot import tags (tags already imports nodes), so the shared
// rules live here, in a package that imports nothing from the graph.
package tagname

import (
	"errors"
	"fmt"
	"strings"
)

// Separator nests a tag under a parent.
const Separator = "/"

// ErrInvalid is the sentinel behind every rejection here, so callers can map a
// bad tag name to a 400 without matching on the message.
var ErrInvalid = errors.New("tagname: invalid tag name")

// Normalize returns the canonical form of a tag name: trimmed, lowercased, and
// with each `/`-separated segment trimmed. Lowercasing is what makes tags a
// matching axis rather than prose — tags.name is UNIQUE, so without it
// `Homelab` and `homelab` would be two subjects that never meet.
//
// It rejects names that would break the nesting convention: empty or
// whitespace-only names, a leading or trailing separator, and empty segments
// (`foo//bar`). Every error wraps ErrInvalid.
func Normalize(name string) (string, error) {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return "", fmt.Errorf("%w: empty", ErrInvalid)
	}

	segments := strings.Split(strings.ToLower(trimmed), Separator)
	for i, seg := range segments {
		seg = strings.TrimSpace(seg)
		if seg == "" {
			return "", fmt.Errorf("%w: %q has an empty segment", ErrInvalid, name)
		}
		segments[i] = seg
	}
	return strings.Join(segments, Separator), nil
}

// Validate reports whether name is a well-formed tag name.
func Validate(name string) error {
	_, err := Normalize(name)
	return err
}

// NormalizeAll normalises every name, preserving order and dropping duplicates
// that only differed by case or padding. It fails on the first invalid name.
func NormalizeAll(names []string) ([]string, error) {
	out := make([]string, 0, len(names))
	seen := make(map[string]struct{}, len(names))
	for _, name := range names {
		normalized, err := Normalize(name)
		if err != nil {
			return nil, err
		}
		if _, dup := seen[normalized]; dup {
			continue
		}
		seen[normalized] = struct{}{}
		out = append(out, normalized)
	}
	return out, nil
}

// ValidateAll returns the first error among names, if any.
func ValidateAll(names []string) error {
	_, err := NormalizeAll(names)
	return err
}

// Segments splits a normalised name into its hierarchy levels.
func Segments(name string) []string {
	return strings.Split(name, Separator)
}

// IsUnder reports whether name is ancestor itself or one of its descendants.
// It is the in-memory twin of the SQL prefix match in tags.NodesUnder, and it
// exists so a caller filtering an already-loaded tag list cannot drift from the
// query semantics: `home` must not match `homelab`.
func IsUnder(name, ancestor string) bool {
	return name == ancestor || strings.HasPrefix(name, ancestor+Separator)
}
